stage/manpages/man1/vmserver.1
.TH vmserver 1  vovida.org 
.\" auto-generated from vmserver\&.xml by docbook2manxml
.SH NAME
vmserver    \- VOCAL voice mail control server  
.SH SYNOPSIS
\fBvmserver\fR
[\fBoptions\fR]
[\fBsip-port\fR]
.SH DESCRIPTION
vmserver controls the voicemail system by controlling a voice mail
user agent (UA) which actually handles the SIP calls\&. vmserver knows about
things like answering the call, which greeting to play, and where to send
the \&.wav file which contains the caller\&'s voice mail message\&.
.SH OPTIONS
.TP 
-d
fork into background as a daemon(NOTE: this is opposite all
other commands)
.TP 
-i
run from inetd
.TP 
-f arg
specify voicemail configuration file
.TP 
-h arg
specify number of VMCP handles
.TP 
-v arg
specify the debugging level
.TP 
-m
run in IVR mode\&. In this mode, vmserver listens on port 9024 instead of 8024\&. the corresponding option to ua is -a\&.
.SH SEE ALSO
\fBvocal\fR(8)
.PP
More information is available from http://www\&.vovida\&.org/<URL:http://www\&.vovida\&.org/>
.SH COPYRIGHT
Copyright 2002-2003, Cisco Systems, Inc\&.
.PP
THE INFORMATION HEREIN IS PROVIDED ON AN "AS IS" BASIS, WITHOUT ANY
WARRANTIES OR REPRESENTATIONS, EXPRESS, IMPLIED OR STATUTORY, INCLUDING
WITHOUT LIMITATION, WARRANTIES OF NONINFRINGEMENT, MERCHANTABILITY OR
FITNESS FOR A PARTICULAR PURPOSE\&.


doc/CHANGELOG.txt
=========================================================================

                          VOCAL CHANGELOG

=========================================================================

This file documents major changes to VOCAL.  Please keep in sync with
the appropriate section of the README.txt .

Changelog Format: yyyy-mm-dd (component/component/...): <report> .
List of components: sip ms ua rs fs util h323 mgcp ps java install loadgen
                    fsvm
----------------------------------------------------------------------

2003-05-07 (sip)

Automatic call timeout if not answered, new TimerThread class tracks calls that haven't been answered and sends a 480 to the calling party and a CANCEL to the called party if the call isn't picked up in a certain amount of time. The timeout is specified in call container.
  
2003-05-07 (ps)

New provisiong system added. The new provisioning system uses a postgres database backend to store most of the provisioning system. See readme files in newprov directory and bug 697. The new provisioning system partially replaces the old one but some things (notably features) require provisioned in both at the moment. The README file in newprov contains the full list of issues and work remaining.


2003-05-07 

Support for G729 VoiceAge library added. See notes in libmedia directory. The code compiles without the library present and uses a 'fake' G729 implementation (which will just produce garbage!). The VoiceAge library can just be placed in the directory and the correct lines commented in in the Makefile to enable it.


New features in 1.5.0:

Provisioning Security

This release includes a number of fixes for provisioning security.  As
a consequence, users upgrading existing installations must upgrade the
provisioning server, HTML and Java GUIs, and all applications
simultaneously.

Users choosing to upgrade must create new files for provisioning
passwords, change their HTML files, and the like.

Please see doc/UPGRADING.txt for (incomplete) information on upgrading
a running VOCAL 1.4 system to VOCAL 1.5.

In addition, TLS support can be enabled at compile time via configure.
Please see README_BUILD.txt for details.

More information about provisioning security can be found in the file
provisioning/README.secprov .

SIPSet

This release also includes the source code to SIPSet.  More
information about SIPSet is available on the VOCAL web site at

	    http://www.vovida.org/applications/downloads/sipset/


CERT Advisory

CERT has released an advisory concerning SIP software, at
href="http://www.cert.org/advisories/CA-2003-06.html. VOCAL 1.5.0
fixes a number of buffer overflow vulnerabilities revealed by these
test cases.

Configured Routing

The RS generates a route set that the request must traverse. The code to
determine the route set and the position of the querying Vocal server in the
route using the top Via in the request are not changed. However, in addition
to return the next hop URI in the contact, the rest of route set, including
the next hop and before any forking, will be embedded with the "?" mechanism
in the Contact URI.  Note that it is not necessary to add the embedded part if
the next hop is the only URI in the rest of the route set. For example, a
common MS for both the UAC and UAS of a dialog.

A Vocal server (e.g. MS) proxies a SIP request to the RS for routing
information if the Route header is not found in the message and the request is
not part of an existing dialog.  If the in-coming SIP request has a Route
header, the proxy forwards the message to the next hop.  If there is no route
and the request is part of an existing dialog, the proxy (dialog stateful)
sends the message to the address and port in the request URI.

The server which gets the 302 (Moved Temporary) response will then extract the
embedded route set and place them in the Route header of the SIP request
before proxying the message to the next hop.  A dialog stateful proxy adds
itself to the Record-Route header and creates an entry in its list of active
dialogs.  The SIP stack sends the message according to the Route header.  Note
that the entries in the route set can be loose or straight routing URIs, but
they don't affect the operation of Configured Route at all.


Loose Routing Support in SIP stack.
Note that Vocal components are still strict routing applications.


2003/03/28

Fixed bugs : 425, 632, 652, 665, 669, 677, 678


New features in 1.5.0:

Provisioning Security

This release includes a number of fixes for provisioning security.  As
a consequence, users upgrading existing installations must upgrade the
provisioning server, HTML and Java GUIs, and all applications
simultaneously.

Users choosing to upgrade must create new files for provisioning
passwords, change their HTML files, and the like.

Please see doc/UPGRADING.txt for (incomplete) information on upgrading
a running VOCAL 1.4 system to VOCAL 1.5.

In addition, TLS support can be enabled at compile time via configure.
Please see README_BUILD.txt for details.

More information about provisioning security can be found in the file
provisioning/README.secprov .

SIPSet

This release also includes the source code to SIPSet.  More
information about SIPSet is available on the VOCAL web site at

	    http://www.vovida.org/applications/downloads/sipset/


ATA Autoprovisioning

ATA Autoprovisioning allows you to plug an ATA into a network running
VOCAL and have the ATA receive the necessary provisioning parameters
to make calls in the system, while simultaneously creating the
appropriate entries in the VOCAL configuration.  Please see
provisioning/auto/README for more information about ATA
autoprovisioning.


2003/02/19
Re-factor the temp file creation in SIP user agent for named pipe to 
address security issues.Support for NAT traversal using NATIP field in 
configuration.


2003/02/19
Re-factor the temp file creation in SIP user agent for named pipe to 

vm/media_server/vmserver.cxx
/* ====================================================================
 * The Vovida Software License, Version 1.0 
 * 
 * Copyright (c) 2000 Vovida Networks, Inc.  All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 
 * 3. The names "VOCAL", "Vovida Open Communication Application Library",
 *    and "Vovida Open Communication Application Library (VOCAL)" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written
 *    permission, please contact vocal@vovida.org.
 *
 * 4. Products derived from this software may not be called "VOCAL", nor
 *    may "VOCAL" appear in their name, without prior written
 *    permission of Vovida Networks, Inc.
 * 
 * THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND
 * NON-INFRINGEMENT ARE DISCLAIMED.  IN NO EVENT SHALL VOVIDA
 * NETWORKS, INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT DAMAGES
 * IN EXCESS OF $1,000, NOR FOR ANY INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 * 
 * ====================================================================
 * 
 * This software consists of voluntary contributions made by Vovida
 * Networks, Inc. and many individuals on behalf of Vovida Networks,
 * Inc.  For more information on Vovida Networks, Inc., please see
 * <http://www.vovida.org/>.
 *
 */

static const char* const main_cpp_Version = 
"$Id: vmserver.cxx,v 1.1 2004/05/01 04:15:38 greear Exp $";
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include "Timer.h"
#include "Waiter.h"

#include "cpLog.h"
#include "LineVMCP.h"
#include "VmSession.h"

#include <signal.h>
#include <unistd.h>

/** This is the main class for voice mail server (vmserver)
  * VM server uses sockets for communication with clients.
  * It listens on port 8024 for requests. 
  */


/** shutdown flag */
bool g_bShutdown = false;
string g_xCfgPath;

extern "C" {
    void catch_child(int p);
    void catch_term(int p);
};

void
catch_child (int p)
{
    /** Child was terminated, get return code. */
    wait (NULL);
    signal (SIGCHLD, catch_child);
}

void
catch_term (int p)
{
    /** Termination signal received. Shutown server. 
     */
    g_bShutdown = true;
    exit(1); 
}

int
daemon()
{
    pid_t pid;
    char* envValString = getenv( "NO_DAEMON" );
    if( envValString )
    {
        if( atoi( envValString ) != 0 )
        {
            cpLog(LOG_DEBUG,"NO_DAEMON environment variable set");
            return 0 ;
        }
    }

    if( ( pid = fork() ) < 0 )
    {
        cpLog(LOG_ERR,"Fork failed!");
        return  -1 ;
    }
    else if( pid != 0 )
    {
        exit ( 0 );
    }
    setsid();
    umask(0);
    close(0);
    return 0;
}


int
main (int argc, char *argv[])
{

    int logPrio=LOG_ERR;
    bool daemonflag=false;
    bool inetdflag=false;
    bool interactive=false;
    int vmcpHandle;
    
    g_xCfgPath="./voicemail.config";

    int c = 0;
	int port = 8024;
  
    while( ( c = getopt( argc, argv, "idf:v:h:m") ) != EOF )
    {
	switch(c)
	{
	    /** To run VM Server as the basis for IVR style applications, interactive 
	     * i.e. 'm' flag is used. This is going to be in future release 
		 * The system uses port 9024 for IVR 
	     */
	case 'm':
	{
	    interactive=true;
		port = 9024;		//Need to be read in by LINEPARSER
	}
	break;

	/** With this option, provide a voicemail configuration file */ 
        case 'f':
 	{
	    g_xCfgPath = optarg;
 	}
	break;
	

	case 'h':
	{
	    if( optarg == 0 ) 
		vmcpHandle=0;
	    else
	    {
		vmcpHandle = atoi(optarg);
	    }
	}
	break;
	 
	/** With 'v' option : specify the Log Priority Level */
	case 'v':
	{
	    if( optarg == 0 ) 
		logPrio = 7;
	    else
	    {
		logPrio = cpLogStrToPriority(optarg);
	    }
	}
	break;

	/** Run The VMServer as a daemon */
	case 'd':
	{
	    daemonflag=true;
	}
	break;

	/** Run The VMServer as a Inet daemon */ 
	case 'i':
	{
	    inetdflag=true;
	}
	break;
      
	}
    }
  
    /** Setup Logging priority */
    cpLogSetLabel ("VM");
    cpLogSetPriority (logPrio);

    Configuration cfg;

    /** Parse configuration */
    cfg.parse (g_xCfgPath);
  
    cpLogSetPriority (logPrio);
    cpLogOpen(cfg.m_xLogFileName.c_str());

    if( daemonflag ) daemon();

    /**  Create sockets */
    Vmcp vmcp;
  
    if( inetdflag )
    {
	vmcp.assign(vmcpHandle,vmcpHandle);
	char logLabel[11];
	sprintf (logLabel, "VM%08X", getpid ());
	cpLogSetLabel (logLabel);
	cpLog (LOG_INFO, "Starting session %08X", getpid ());

	/** Close waiting socket. We don't need it in this process */
	/** Start session */
	VmSession session;
	session.Start(&vmcp,interactive);
	cpLog (LOG_INFO, "End of session");
	return 3;

    }
    /** Catch SIGCHLD signal to get rid of zombies */
    signal (SIGCHLD, catch_child);

    /** Catch all termination signals to shutdown gracefully */
    signal (SIGTERM, catch_term);
    signal (SIGHUP, catch_term);
    signal (SIGKILL, catch_term);
    signal (SIGABRT, catch_term);
    signal (SIGSTOP, catch_term);
    signal (SIGTSTP, catch_term);
    signal (SIGINT, catch_term);
    signal (SIGQUIT, catch_term);


    /** Open a socket 
     *  Voice Mail server always listens for the requests on port 8024 
     */
    int sfd;
    if( (sfd=vmcp.wait(port))==-1 )
    {
	cpLog(LOG_ERR,"Error binding socket %d",sfd );
	exit(3);
    }

    while (!g_bShutdown)
    {
	int fd;
	cpLog(LOG_DEBUG,"Waiting for connection");
	fd=vmcp.accept();
	if( fd ==-1 )
	{
	    cpLog(LOG_ERR,"Error accepting socket" );
	    exit(3);
	}

	if (g_bShutdown)
	{
	    break;
	}
	
	cpLog (LOG_DEBUG, "Connected");

	/** Ok, we have connected client. */
	/** Let's start a new process for it. */
	switch (fork ())
	{
	case 0:
	{
	    close(vmcp.getListenSocket());
	    /** set log label for a new process */
	    char logLabel[11];
	    sprintf (logLabel, "VM%08X", getpid ());
	    cpLogSetLabel (logLabel);
	    cpLog (LOG_INFO, "Starting session", getpid ());

	    /** Start session */
	    VmSession session;
	    session.Start(&vmcp,interactive);
	    cpLog (LOG_INFO, "End of session");
	    return 3;
	}
	default:
	    /** Close connected socket. We don't need it in this process. */
	    /** Continue loop */
	    cpLog(LOG_DEBUG, "In parent: closing");
	    close(fd);
	    break;
	}
    }

    cpLog (LOG_INFO, "Server shutdown.");


    return EXIT_SUCCESS;
}
